x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+            "Invalid value [" + jwkSetPath + "] for setting " + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+        );
+        settingsException.addSuppressed(urlException);
+        settingsException.addSuppressed(pathException);
+        throw settingsException;
+    }
+
+    private void ensureExpectedValueForInitialized(final boolean expectedValue) {
+        if (this.initialized != expectedValue) {
+            if (expectedValue) {
+                throw new IllegalStateException("Realm has not been initialized");
+            } else {
+                throw new IllegalStateException("Realm has already been initialized");
+            }
+        }
+    }
+
+    @Override
+    public void initialize(final Iterable<Realm> allRealms, final XPackLicenseState xpackLicenseState) {
+        this.ensureExpectedValueForInitialized(false);
+        // extract list of realms referenced by super.config.settings() value for DelegatedAuthorizationSettings.AUTHZ_REALMS
+        this.delegatedAuthorizationSupport = new DelegatedAuthorizationSupport(allRealms, super.config, xpackLicenseState);
+        this.initialized = true;
+    }
+
+    @Override
+    public boolean supports(final AuthenticationToken jwtAuthenticationToken) {
+        this.ensureExpectedValueForInitialized(true);
+        return (jwtAuthenticationToken instanceof JwtAuthenticationToken);
+    }
+
+    @Override
+    public AuthenticationToken token(final ThreadContext threadContext) {
+        this.ensureExpectedValueForInitialized(true);
+        final SecureString authorizationParameterValue = JwtRealm.getHeaderSchemeParameters(
+            threadContext,
+            JwtRealmSettings.HEADER_ENDUSER_AUTHORIZATION,
+            JwtRealmSettings.HEADER_ENDUSER_AUTHORIZATION_SCHEME,
+            false
+        );
+        if (authorizationParameterValue == null) {
+            return null; // Could not find non-empty SchemeParameters in HTTP header "Authorization: Bearer <SchemeParameters>"
+        }
+
+        // Get all other possible parameters. A different JWT realm may do the actual authentication.
+        final SecureString clientAuthorizationSharedSecretValue = JwtRealm.getHeaderSchemeParameters(
+            threadContext,
+            JwtRealmSettings.HEADER_CLIENT_AUTHORIZATION,
+            JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET,
+            true
+        );
+
+        return new JwtAuthenticationToken(authorizationParameterValue, clientAuthorizationSharedSecretValue);
+    }
+
+    @Override
+    public void authenticate(final AuthenticationToken authenticationToken, final ActionListener<AuthenticationResult<User>> listener) {
+        this.ensureExpectedValueForInitialized(true);
+        if (authenticationToken instanceof JwtAuthenticationToken jwtAuthenticationToken) {
+            LOGGER.trace("Realm [{}] supports JwtAuthenticationToken.", super.name());
```

Comment:
I don't think this is a very useful trace log.
It might make sense to have something like:
 ```suggestion
             LOGGER.trace("Attempt to perform authentication of token [{}] with realm [{}]", jwtAuthenticationToken.getSerializedJwt(), super.name());
```
That helps trace the code as it runs so if there are cases where it's failing unexpected we can track where it gets to.

Reply:
I can tweak the message, but is it OK to include the JWT value? In JwtAuthenticationToken to had a comment to avoid returning the JWT value in principal() or credential(), because those return values go into audit logging. I will update the message but not include the JWT value. If you think it is OK to include it, I can add it next time.

Reply:
Done